src/fsharp/Optimizer.fs

```diff
@@ -2901,6 +2901,55 @@ and OptimizeApplication cenv env (f0, f0ty, tyargs, args, m) =
                    MightMakeCriticalTailcall = mayBeCriticalTailcall
                    Info=ValueOfExpr newExpr }
 
+and TryApplicationToLambda (v: Val) cenv env topValInfo (f, fty, tyargs, args: Exprs, m) ety =
+    let g = cenv.g
+
+    match f with
+    // For partial applications, try to wrap it in a lambda as they have better optimization strategies.
+    | Expr.Val (vref, _, _) when cenv.settings.localOpt() && not v.IsCompiledAsTopLevel && 
+                                 not v.IsMutable && not vref.IsMutable && 
+                                 not vref.IsTypeFunction && isFunTy g ety ->
+        match vref.ValReprInfo with
+        | Some valInfo ->
+            // We check the number of curried arguments with the applied arguments to ensure that it's safe to turn
+            //     this app expr into a lambda one.
+            // Example:
+            (*
+                let test a b c d e f g : int -> unit =
+                    printfn "hello"
+                    System.Console.WriteLine
+
+                let test2 a b c d e f g xs =
+                    let action = test a b c d e f g
+                    printfn "world"
+                    xs |> Array.iter action
+            *)
+            // In the example, we should not make a lambda around "test" in "test2" because "test" is fully applied but returns
+            //     a partial application.
+
+            // The currently passed arguments must not have any effects.
+            let argTys, retTy = stripFunTy g ety
+            if valInfo.NumCurriedArgs = args.Length + argTys.Length && not (ExprsHaveEffect g args) then
```

Comment:
I think we can't exactly convert an application to a lambda when all args simply don't have effects.
The args must be values that are not mutable.